### PR TITLE
Parameterized queries

### PR DESCRIPTION
**Pull request by PrashamTrivedi**
Couple of points

- I tried to re-write the tests such a way that all pass. But somehow, couple of tests are passing (and not passing) at will. They are mostly Rx or Async tests. I will look into them later this week once I got free. Meanwhile If you or anyone can point out where I went wrong, it would be most helpful

- Overall, my addition of tests is 
    a) To show how a call is changed from current behavior (or in other words how parameterised queries should be written)
    b) To ensure both methods produce same data (This is where I fail a bit)

Couple of other useful additions

- Proguard consumer files: Proguard rules will automatically be added once this dependency is added.

- Compiled against Latest AS Canary, Kotlin 1.2, Latest Realm plugin and SDK 27

### CODE REVIEW

File: library-base/build.gradle [hunk 1]
```diff
@@ -34,9 +34,9 @@ android {
 }
 
 dependencies {
-    provided 'io.reactivex.rxjava2:rxjava:2.1.4'
-    provided 'io.reactivex.rxjava2:rxandroid:2.0.1'
-    compile fileTree(dir: 'libs', include: ['*.jar'])
+    compileOnly 'io.reactivex.rxjava2:rxjava:2.1.4'
```
Comment (PrashamTrivedi) [inline library-base/build.gradle:37]: Updated to new dependency configurations as guided [here](https://developer.android.com/studio/build/gradle-plugin-3-0-0-migration.html#new_configurations)

**Comment by vicpinm** [comment]
Hi @PrashamTrivedi, you are right. I've executed the test and there seems to be some error with the rx subscription that causes the test to fail. I will try to see what is happening in the coming days. 

Thanks you so much for your contribution.

**Comment by vicpinm** [comment]
@PrashamTrivedi, the problem seems to come from lines like this one: 

assertThat(disposable?.isDisposed ?: false).isTrue()

This is the assertion is failing, yo can see it if you print the error in all the subscribers, implementing the onError callback:

    java.lang.AssertionError: The subject was expected to be true, but was false
                                                 at  com.vicpin.krealmextensions.KRealmExtensionsRxTests.testQuerySortedAsSingleParameterized(KRealmExtensionsRxTests.kt:280)

The test makes the assumption that the disposable must be disposed automatically after emiting the single item, in methods like testQuerySortedAsSingleParameterized. I'm not sure why isDisposed is false at that point because it is a single observable which should terminate automatically after emmiting its single item. I will continue investigating this in the coming days.

**Comment by vicpinm** [comment]
@PrashamTrivedi I know where the problem is. You method querySortedAsSingle is returning a Flowable instead of a Single. That's the reason the subscription is not disposed after emmiting the result.

**Comment by PrashamTrivedi** [comment]
> You method querySortedAsSingle is returning a Flowable instead of a Single. 

Missed it. Updating the code and re-running the tests right now. 

Updating PR, with new method to properly return singles and adding @PublishedApi annotation in `getLooper()` method so it can be used within inline functions

Edit: This PR has all tests passed, thanks for guiding.

File: library-base/src/main/java/com/vicpin/krealmextensions/RealmExtensionsSingle.kt [hunk 2]
```diff
@@ -3,28 +3,64 @@ package com.vicpin.krealmextensions
...
 fun <T : RealmModel> T.querySortedAsSingle(fieldName : List<String>, order : List<Sort>, query: Query<T>? = null) = performQuery(fieldName, order, query)
+inline fun <reified T : RealmModel> querySortedAsSingle(fieldName : List<String>, order : List<Sort>, noinline query: Query<T>? = null) = performQueryAndReturnSingle(fieldName, order, query)
 
 /**
  * Query for sorted entities and observe changes returning a Single.
  */
 fun <T : RealmModel> T.querySortedAsSingle(fieldName : String, order : Sort, query: Query<T>? = null) = performQuery(listOf(fieldName), listOf(order), query)
+inline fun <reified T : RealmModel> querySortedAsSingle(fieldName : String, order : Sort, noinline query: Query<T>? = null) = performQueryAndReturnSingle(listOf(fieldName), listOf(order), query)
+
+inline fun <reified T:RealmModel> performQueryAndReturnSingle(fieldName: List<String>? = null, order: List<Sort>? = null, noinline query: Query<T>? = null):Single<List<T>>{
```
Comment (PrashamTrivedi) [inline library-base/src/main/java/com/vicpin/krealmextensions/RealmExtensionsSingle.kt:36]: Added new method which returns single. Effectively solving the issue of not passing tests.

**Comment by vicpinm** [comment]
@PrashamTrivedi I have merged your changes into master. I've made some refactors, like creating new test clases only for parameterized methods. I will upload a new version to maven as soon as I review the other PR. 
Thanks a lot for your contribution!

**Pull request closed without merging**